mlflow/store/artifact/databricks_artifact_repo.py

```diff
@@ -386,9 +389,41 @@ def _azure_upload_file(self, credentials, local_file, artifact_file_path, get_cr
                     size=MLFLOW_MULTIPART_UPLOAD_CHUNK_SIZE.get(),
                     get_credentials=get_credentials,
                 )
-                futures[future] = index
 
-            results, errors = _complete_futures(futures, local_file)
+            futures = {
+                self.chunk_thread_pool.submit(job, index): index for index in range(num_chunks)
+            }
+
+            errors = []
+            results = {}
+            timeout = MLFLOW_MULTIPART_CHUNK_UPLOAD_TIMEOUT.get()
+            retries = MLFLOW_MULTIPART_CHUNK_UPLOAD_MAX_RETRIES.get()
+            for i in range(retries):
+                new_futures = {}
+                for f in futures:
+                    try:
+                        res = f.result(timeout=timeout)
+                        index = futures[f]
+                        results[index] = res
+                    except concurrent.futures.TimeoutError as e:
+                        if i == retries - 1:
+                            index = futures[f]
+                            new_futures[self.chunk_thread_pool.submit(job, index)] = index
+                            continue
+
+                        errors.append(e)
+                    except Exception as e:
+                        errors.append(e)
+
+                if not new_futures:
+                    break
+
+                # Fail fast if we encounter any unexpected errors
+                if errors:
+                    break
+
+                futures = new_futures
```

Comment:
@BenWilson2 @smurching Is it possible to get an early review to make sure we're on the same page?

Reply:
Directionally this is exactly what I think we need! Let's proceed with mocked failure tests that simulate multiple retries and retry loop exhaustion handling, and I think we'll be good to go. Thanks @harupy !!